mavenembedder-workitem/src/main/java/org/jbpm/process/workitem/mavenembedder/MavenEmbedderWorkItemHandler.java

```diff
@@ -0,0 +1,121 @@
+import org.jbpm.process.workitem.core.util.WidParameter;
+import org.jbpm.process.workitem.core.util.WidResult;
+import org.jbpm.process.workitem.core.util.service.WidAction;
+import org.jbpm.process.workitem.core.util.service.WidService;
+import org.kie.api.runtime.process.WorkItem;
+import org.kie.api.runtime.process.WorkItemManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Wid(widfile = "MavenEmbedderDefinitions.wid", name = "MavenEmbedder",
+        displayName = "MavenEmbedder",
+        defaultHandler = "mvel: new org.jbpm.process.workitem.mavenembedder.MavenEmbedderWorkItemHandler()",
+        documentation = "${artifactId}/index.html",
+        parameters = {
+                @WidParameter(name = "Commands", required = true),
+                @WidParameter(name = "WorkDirectory", required = true),
+                @WidParameter(name = "ProjectRoot")
+        },
+        results = {
+                @WidResult(name = "MavenResults")
+        },
+        mavenDepends = {
+                @WidMavenDepends(group = "${groupId}", artifact = "${artifactId}", version = "${version}")
+        },
+        serviceInfo = @WidService(category = "${name}", description = "${description}",
+                keywords = "maven,mvn,build,execute,pom,project,intall",
+                action = @WidAction(title = "Execute Maven commands")
+        ))
+public class MavenEmbedderWorkItemHandler extends AbstractLogOrThrowWorkItemHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(MavenEmbedderWorkItemHandler.class);
+    private static final String RESULTS_VALUES = "MavenResults";
+
+    public void executeWorkItem(WorkItem workItem,
+                                WorkItemManager workItemManager) {
+
+        try {
+
+            RequiredParameterValidator.validate(this.getClass(),
+                                                workItem);
+
+            String commands = (String) workItem.getParameter("Commands");
+            String workDir = (String) workItem.getParameter("WorkDirectory");
+            String projectRoot = (String) workItem.getParameter("ProjectRoot");
+            if(projectRoot == null || projectRoot.length() < 1) {
+                projectRoot = "$M2_HOME";
+            }
+
+            Map<String, Object> results = new HashMap<String, Object>();
+
+            ByteArrayOutputStream baosOut = new ByteArrayOutputStream();
+            ByteArrayOutputStream baosErr = new ByteArrayOutputStream();
+
+            PrintStream cliOut = new PrintStream(baosOut,
+                                                 true);
+            PrintStream cliErr = new PrintStream(baosErr,
+                                                 true);
+
+            MavenCli cli = new MavenCli();
+
+            String origMultiModuleProjectRoot = System.getProperty("maven.multiModuleProjectDirectory");
+            System.setProperty("maven.multiModuleProjectDirectory", projectRoot);
```

Comment:
what these two lines do?

Reply:
since maven 3.3.1 maven.multiModuleProjectDirectory prop was added. It used to find the location of .mvn. Maven embedder needs it set. The intent of these lines is to first get any existing setting that might be in the execution env (orig) and the set it to whats passed to the handler via params, then at the end reset it to what it was originally set to we dont overwrite anything after execution of the workitem.